File: src/lab2.js
async function lex(body) {
    let text = ""
    let in_angle = false
    for (let c of body) {
        if (c == "<")
            in_angle = true
        else if (c == ">")
            in_angle = false
        else if (!in_angle)
            text += c
        await potentialBreakpointLex(text);
    }
    return text
}

let [WIDTH, HEIGHT] = [800, 600]
let [HSTEP, VSTEP] = [10, 18]

let SCROLL_STEP = 100

async function layout(text) {
    let display_list = []
    let [x, y] = [HSTEP, VSTEP]
    for (let c of text) {
        display_list.push([x, y, c])
        x += HSTEP
        if (x >= WIDTH - HSTEP) {
            y += VSTEP
            x = HSTEP
        }
        await potentialBreakpointLayout(display_list)
    }
    return display_list
}

class Browser {
  constructor(canvasElement) {
    this.canvasElement = canvasElement;
    const rectangle = canvasElement.getBoundingClientRect();
    canvasElement.width = rectangle.width * devicePixelRatio;
    canvasElement.height = rectangle.height * devicePixelRatio;
    this.canvasContext = canvasElement.getContext('2d');
    this.scroll = 0;
  }

  async load(body) {
    this.clearCanvas();
    let text = await lex(body);
    this.display_list = await layout(text);
    await this.render();
  }

  clearCanvas() {
    this.canvasContext.clearRect(0, 0, this.canvasElement.width,
        this.canvasElement.height);
  }

  async render() {
    let count = 0;
    this.canvasContext.font = "normal normal 16px sans-serif";
    console.log(this.canvasContext.font)
    for (let entry of this.display_list) {
      let [x, y, c] = entry;
      if (y > this.scroll + HEIGHT)
        continue;
      if (y + VSTEP < this.scroll)
        continue;
      this.canvasContext.fillText(c, x, y - this.scroll);
      await potentialBreakpointRender(count++);
    }
  }

  scrolldown() {
    this.scroll += SCROLL_STEP;
    this.render()
  }
}

File: src/lab2.py
"""
This file compiles the code in Web Browser Engineering,
up to and including Chapter 2 (Drawing to the Screen),
without exercises.
"""

import wbetools
import socket
import ssl
import tkinter
from lab1 import URL

def lex(body):
    text = ""
    in_tag = False
    for c in body:
        if c == "<":
            in_tag = True
        elif c == ">":
            in_tag = False
        elif not in_tag:
            text += c
        wbetools.record("lex", text)
    return text

WIDTH, HEIGHT = 800, 600
HSTEP, VSTEP = 13, 18

SCROLL_STEP = 100

def layout(text):
    display_list = []
    cursor_x, cursor_y = HSTEP, VSTEP
    for c in text:
        display_list.append((cursor_x, cursor_y, c))
        cursor_x += HSTEP
        if cursor_x >= WIDTH - HSTEP:
            cursor_y += VSTEP
            cursor_x = HSTEP
        wbetools.record("layout", display_list)
    return display_list

class Browser:
    def __init__(self):
        self.window = tkinter.Tk()
        self.canvas = tkinter.Canvas(
            self.window,
            width=WIDTH,
            height=HEIGHT
        )

        self.canvas.pack()

        self.scroll = 0
        self.window.bind("<Down>", self.scrolldown)

    def load(self, url):
        body = url.request()
        text = lex(body)
        self.display_list = layout(text)
        self.draw()

    def draw(self):
        self.canvas.delete("all")
        for x, y, c in self.display_list:
            wbetools.record("draw")
            if y > self.scroll + HEIGHT: continue
            if y + VSTEP < self.scroll: continue
            self.canvas.create_text(x, y - self.scroll, text=c)

    def scrolldown(self, e):
        self.scroll += SCROLL_STEP
        self.draw()

if __name__ == "__main__":
    import sys

    Browser().load(URL(sys.argv[1]))
    tkinter.mainloop()
